Array collections with modifiers (#62)
yiisoft/arrays: README.md

[\ArrayAccess](https://www.php.net/manual/class.arrayaccess) and
[\Countable](https://www.php.net/manualn/class.countable.php);
- `ArrayableInterface` and `ArrayableTrait` for use in classes who want to support customizable representation of their instances.

> - `ArrayCollection` that allow use arrays with modifiers.

Reviewer: ```suggestion
- `ArrayCollection` that allow to apply modifiers to arrays.
```

$data = $car->toArray(['type', 'color']); // ['type' => 'Crossover', 'color' => 'Red']
```

## ArrayCollection usage

> `ArrayCollection` is array with modifiers. When you get array value or whole array from collection

Reviewer: ```suggestion
`ArrayCollection` is array wrapper that allows specifying modifiers. When you get array value or whole array from the collection
```

> modifiers are applied and get changed data.

Reviewer: ```suggestion
modifiers are applied first so you get modified data.
```

> When merging collections (use `ArrayHelper::merge` or `$collection->mergeWith()`) separately combines original arrays and modifiers.

Reviewer: ```suggestion
When merging collections using `ArrayHelper::merge()` or `$collection->mergeWith()` original arrays and modifers are merged separately.
```

Example of use:

```php
$collection = new ArrayCollection(
    [
        'name' => 'Yii',
        'version' => '2.0',
        'tag' => 'php',
    ],
    new SaveOrder(),
);

$update = new ArrayCollection(
    [
        'description' => 'PHP framework',
        'version' => '3.0',
    ],
    new UnsetValue('tag'),
    new MoveValueBeforeKey('description', 'version')
);

$update2 = [
    'license' => 'BSD',
];

// [
//    'name' => 'Yii',
//    'description' => 'PHP framework',
//    'version' => '3.0',
//    'license' => 'BSD',
// ]
$result = ArrayHelper::merge($collection, $update, $update2);
```

### Modifiers

> 

Reviewer: ```suggestion

Modifiers are specified as extra constructor elements of `ArrayCollection`:

```php
$dataSet = new ArrayCollection($data, new Modifier1(), new Modifier2());
```

The following modifiers are available.

```

#### MoveValueBeforeKey

```php
new MoveValueBeforeKey('key', 'beforeKey');
```

> Move element with key `key` before key `beforeKey`.

Reviewer: ```suggestion
Move element with a key `key` before an element with `beforeKey` key.
```

#### RemoveAllKeys

```php
new RemoveAllKeys();
```

> Indexes an array numerically.

Reviewer: ```suggestion
Re-indexes an array numerically, i.e. removes all information about array keys.
```

#### ReplaceValue

```php
new ReplaceValue('key');
```

The modifier allows to mark an array element from the collection it is applied to,
as the element to be processed in a special way on merge.
 

> - In case there are elements with the same keys in previous arrays, they will be replaced
>   with a value from the current array.
> 
> - If there are elements with the same keys in next arrays, they will replace current array value.

Reviewer: ```suggestion
- In case there are elements with the same keys in previous arrays, they will be replaced
  with a value from the current array.
- If there are elements with the same keys in next arrays, they will replace current array value.
```

$data = $car->toArray(['type', 'color']); // ['type' => 'Crossover', 'color' => 'Red']
```

## ArrayCollection usage

`ArrayCollection` is array wrapper that allows specifying modifiers. When you get array value or whole array
from the collection modifiers are applied first so you get modified data.

When merging collections using `ArrayHelper::merge()` or `$collection->mergeWith()` original arrays

> and modifers are merged separately.

Reviewer 2: ```suggestion
and modifiers are merged separately.
```
